File: Rubyとか/14_BinarySearchAlgorithm_MinMaxDivision.rb
# you can write to stdout for debugging purposes, e.g.
# puts "this is a debug message"


def BinarySearch(arr, k)
  # Mの範囲で二分探索
	left = 0 
	right = 10 ** 12 #条件内の最大和 log2(10**12)≒33 なのでNlogNなら間に合う

  # 探索済みについて(最終的に振動するのでその判定)
  hash = {}

	while left <= right 
		mid = (left + right) / 2 
		if check(arr, k, mid)
      return mid if hash.has_key?(mid)
      hash[mid] = true 
      right = mid - 1
    else
			left = mid + 1
		end
    #puts "left:#{left} right:#{right} mid:#{mid} check:#{check(arr, k, mid)}"
	end
  
  return hash.keys.min
end

# arrをk分割する際に各ブロックの和をs以下にできるか
def check(arr, k, s)
  return false if arr.max > s 

  sum = 0
  cnt = 0 
  ind = 0 
  while ind < arr.length 
    sum += arr[ind]
    if sum > s 
      sum = 0
      cnt += 1
      next 
    end
    ind += 1
  end
  return cnt < k 
end

def solution(k, m, a)
  # write your code in Ruby 2.2
 
  return BinarySearch(a, k)
  # p check(a, k, 5)
  # return -1
end

File: Python/14_BinarySearchAlgorithm_MinMaxDivision.py
# you can write to stdout for debugging purposes, e.g.
# print("this is a debug message")

def solution(K, M, A):
    # write your code in Python 3.6

	# 考え方を逆転させる
	# 通常 → どのように分割すれば合計を最小にできるか(分割する位置を探索する)
	# 今回 → 各ブロックの合計がS以下になるように好きな数で分割する
	#　　　　このとき，分割されたブロック数がK以下であればSは条件を満たす
	#　　　　条件を満たすSの内，最小のものを探す(分割する基準となる合計値を探索する)

	#　　　　Sが条件を満たしている場合，Sよりも大きい数値は必ず条件を満たす(調べる必要がない)
	#　　　　逆にSが条件を満たしていない場合，Sよりも小さい数値は必ず条件を満たさない(調べる必要がない)
	#　　　　↑なので二分探索で十分

    # 二分探索でK分割以下にできるSUMを探していく
    ans = 10 ** 12
    left = 0 
    right = 10 ** 12
    # 各要素最大10^6，要素数最大10^6, Log2(10^12)=39ぐらいなので Nlog(10^12) で間に合う

    while left <= right:
        mid = (left + right) // 2
        res = check(K, A, mid)
        if res == True:
            right = mid - 1
        else:
            left = mid + 1
  
        if res == True: ans = min(ans, mid)
    # end of while 

    return ans
# end of func


# ブロックの合計がS以下になるように分割した際に，
# ブロック数をK以下にできるか
def check(K, A, S):
    blockNum = 0
    sum_ = 0
    i = 0 
    while i < len(A):
        sum_ += A[i]
        if sum_ > S:
            sum_ = 0
            blockNum += 1
            if blockNum >= K: break 
            continue  
        
        i += 1
    # end of while
    return blockNum < K
# end of func
